File: tests/unit/email-otp.test.ts
import { describe, expect, it } from 'vitest';
import {
  emailOtpAddressDecoder,
  emailOtpChallengeIdDecoder,
  emailOtpCodeDecoder,
  emailOtpDigestDecoder,
} from '@/lib/domain/email-otp';
import {
  EMAIL_OTP_MAX_FAILED_ATTEMPTS,
  EMAIL_OTP_RATE_LIMIT_WINDOW_SECONDS,
  EMAIL_OTP_RESEND_INTERVAL_SECONDS,
  EMAIL_OTP_TTL_SECONDS,
  createEmailOtpChallenge,
  decideEmailOtpIdentityResolution,
  emailOtpChallengeDecoder,
  invalidateEmailOtpDelivery,
  reserveEmailOtpRateLimit,
  resendEmailOtpChallenge,
  verifyEmailOtpChallenge,
  type EmailOtpChallenge,
  type EmailOtpRateLimitState,
} from '@/server/core/email-otp';
import {
  emailOtpFixture,
  emailOtpIdentityRecord,
  fixturePendingEmailOtp,
} from '@/tests/fixtures/email-otp';
import { sessionFixtureIds } from '@/tests/fixtures/session';

describe('Email OTP boundary codecs', () => {
  it('normalizes bounded ASCII addresses and accepts only eight digits', () => {
    const address = emailOtpAddressDecoder.decode('Person@Example.COM');
    expect(address).toEqual({ ok: true, value: 'Person@example.com' });
    expect(emailOtpAddressDecoder.decode('person..x@example.com').ok).toBe(
      false,
    );
    expect(emailOtpAddressDecoder.decode('person@example.com ').ok).toBe(false);
    expect(emailOtpCodeDecoder.decode('12345678').ok).toBe(true);
    for (const candidate of ['1234567', '123456789', '1234abcd']) {
      expect(emailOtpCodeDecoder.decode(candidate).ok).toBe(false);
    }
    expect(emailOtpChallengeIdDecoder.decode('not-a-uuid').ok).toBe(false);
    expect(emailOtpDigestDecoder.decode(`${'A'.repeat(42)}B`).ok).toBe(false);
  });

  it('rejects malformed persisted challenge timelines and unknown fields', () => {
    expect(emailOtpChallengeDecoder.decode(fixturePendingEmailOtp()).ok).toBe(
      true,
    );
    expect(
      emailOtpChallengeDecoder.decode({
        ...fixturePendingEmailOtp(),
        plaintextCode: emailOtpFixture.code,
      }).ok,
    ).toBe(false);
    expect(
      emailOtpChallengeDecoder.decode(
        fixturePendingEmailOtp({ expiresAtEpochSeconds: 1_601 }),
      ).ok,
    ).toBe(false);
    expect(
      emailOtpChallengeDecoder.decode(
        fixturePendingEmailOtp({ failedAttempts: 5 }),
      ).ok,
    ).toBe(false);
    expect(
      emailOtpChallengeDecoder.decode(
        fixturePendingEmailOtp({ failedAttempts: 1, version: 1 }),
      ).ok,
    ).toBe(false);
  });
});

describe('Email OTP challenge policy', () => {
  it('creates a ten-minute, single-purpose challenge without plaintext code', () => {
    const decision = createEmailOtpChallenge({
      challengeId: emailOtpFixture.challengeId,
      address: emailOtpFixture.address,
      digest: emailOtpFixture.digest,
      salt: emailOtpFixture.salt,
      purpose: { kind: 'sign-in' },
      nowEpochSeconds: 1_000,
    });
    expect(decision.kind).toBe('created');
    if (decision.kind !== 'created') return;
    expect(decision.challenge.expiresAtEpochSeconds).toBe(
      1_000 + EMAIL_OTP_TTL_SECONDS,
    );
    expect(JSON.stringify(decision.challenge)).not.toContain(
      emailOtpFixture.code,
    );
    expect(
      createEmailOtpChallenge({
        challengeId: emailOtpFixture.challengeId,
        address: emailOtpFixture.address,
        digest: emailOtpFixture.digest,
        salt: emailOtpFixture.salt,
        purpose: { kind: 'sign-in' },
        nowEpochSeconds: Number.MAX_SAFE_INTEGER,
      }),
    ).toEqual({ kind: 'rejected', reason: 'invalid-clock' });
  });

  it('consumes a correct code once and invalidates an expired challenge', () => {
    const challenge = fixturePendingEmailOtp();
    const verified = verifyEmailOtpChallenge({
      challenge,
      digestMatches: true,
      nowEpochSeconds: 1_500,
    });
    expect(verified.kind).toBe('verified');
    if (verified.kind !== 'verified') return;
    expect(verified.challenge.kind).toBe('consumed');
    expect(
      verifyEmailOtpChallenge({
        challenge: verified.challenge,
        digestMatches: true,
        nowEpochSeconds: 1_501,
      }),
    ).toEqual({ kind: 'rejected', reason: 'not-pending' });

    const expired = verifyEmailOtpChallenge({
      challenge,
      digestMatches: true,
      nowEpochSeconds: 1_600,
    });
    expect(expired).toMatchObject({
      kind: 'rejected',
      reason: 'expired',
      challenge: { kind: 'invalidated', reason: 'expired' },
    });
  });

  it('locks after five failures and never grants a sixth attempt', () => {
    let challenge: EmailOtpChallenge = fixturePendingEmailOtp();
    for (
      let attempt = 1;
      attempt <= EMAIL_OTP_MAX_FAILED_ATTEMPTS;
      attempt += 1
    ) {
      const decision = verifyEmailOtpChallenge({
        challenge,
        digestMatches: false,
        nowEpochSeconds: 1_000 + attempt,
      });
      expect(decision.kind).toBe('rejected');
      expect(decision.challenge?.failedAttempts).toBe(attempt);
      if (decision.challenge) challenge = decision.challenge;
    }
    expect(challenge.kind).toBe('locked');
    expect(
      verifyEmailOtpChallenge({
        challenge,
        digestMatches: true,
        nowEpochSeconds: 1_100,
      }),
    ).toEqual({ kind: 'rejected', reason: 'not-pending' });
  });

  it('enforces resend cooldown and send cap without resetting failures or expiry', () => {
    const challenge = fixturePendingEmailOtp({ failedAttempts: 2 });
    expect(
      resendEmailOtpChallenge({
        challenge,
        digest: emailOtpFixture.digest,
        salt: emailOtpFixture.otherSalt,
        nowEpochSeconds: 1_000 + EMAIL_OTP_RESEND_INTERVAL_SECONDS - 1,
      }),
    ).toEqual({ kind: 'rejected', reason: 'too-soon' });
    const resent = resendEmailOtpChallenge({
      challenge,
      digest: emailOtpFixture.digest,
      salt: emailOtpFixture.otherSalt,
      nowEpochSeconds: 1_000 + EMAIL_OTP_RESEND_INTERVAL_SECONDS,
    });
    expect(resent).toMatchObject({
      kind: 'resent',
      challenge: {
        failedAttempts: 2,
        sendCount: 2,
        expiresAtEpochSeconds: 1_600,
      },
    });
    expect(
      resendEmailOtpChallenge({
        challenge: fixturePendingEmailOtp({
          sendCount: 3,
          lastSentAtEpochSeconds: 1_100,
        }),
        digest: emailOtpFixture.digest,
        salt: emailOtpFixture.otherSalt,
        nowEpochSeconds: 1_200,
      }),
    ).toEqual({ kind: 'rejected', reason: 'send-limit' });
  });

  it('invalidates a persisted challenge when delivery fails', () => {
    expect(
      invalidateEmailOtpDelivery(fixturePendingEmailOtp(), 1_001),
    ).toMatchObject({
      kind: 'invalidated',
      reason: 'delivery-failed',
      version: 2,
    });
  });
});

describe('Email OTP abuse and identity policy', () => {
  function rateState(counts: {
    readonly address: number;
    readonly network: number;
    readonly account?: number;
  }): EmailOtpRateLimitState {
    const bucket = (key: typeof emailOtpFixture.addressKey, count: number) => ({
      key,
      count,
      windowStartedAtEpochSeconds: 1_000,
    });
    return {
      address: bucket(emailOtpFixture.addressKey, counts.address),
      network: bucket(emailOtpFixture.networkKey, counts.network),
      ...(counts.account === undefined
        ? {}
        : { account: bucket(emailOtpFixture.accountKey, counts.account) }),
    };
  }

  it('reserves all abuse dimensions atomically and resets only after the window', () => {
    expect(
      reserveEmailOtpRateLimit(rateState({ address: 4, network: 4 }), 1_100),
    ).toMatchObject({
      kind: 'reserved',
      state: { address: { count: 5 }, network: { count: 5 } },
    });
    expect(
      reserveEmailOtpRateLimit(rateState({ address: 5, network: 1 }), 1_100),
    ).toEqual({ kind: 'rejected', reason: 'limited' });
    expect(
      reserveEmailOtpRateLimit(
        rateState({ address: 5, network: 30, account: 5 }),
        1_000 + EMAIL_OTP_RATE_LIMIT_WINDOW_SECONDS,
      ),
    ).toMatchObject({
      kind: 'reserved',
      state: {
        address: { count: 1 },
        network: { count: 1 },
        account: { count: 1 },
      },
    });
    expect(
      reserveEmailOtpRateLimit(
        {
          ...rateState({ address: 0, network: 0 }),
          address: {
            ...rateState({ address: 0, network: 0 }).address,
            windowStartedAtEpochSeconds: 1_001,
          },
        },
        1_000,
      ),
    ).toEqual({ kind: 'rejected', reason: 'invalid-clock' });
  });

  it('never auto-links a Google/contact-email collision', () => {
    expect(
      decideEmailOtpIdentityResolution({
        purpose: { kind: 'sign-in' },
        address: emailOtpFixture.address,
        verifiedContactAccountId: sessionFixtureIds.accountId,
      }),
    ).toEqual({ kind: 'rejected', reason: 'email-collision' });
    expect(
      decideEmailOtpIdentityResolution({
        purpose: { kind: 'sign-in' },
        address: emailOtpFixture.address,
        existingIdentity: emailOtpIdentityRecord,
      }),
    ).toEqual({
      kind: 'authenticate-existing',
      identity: emailOtpIdentityRecord,
    });
    expect(
      decideEmailOtpIdentityResolution({
        purpose: { kind: 'link', accountId: sessionFixtureIds.otherAccountId },
        address: emailOtpFixture.address,
        existingIdentity: emailOtpIdentityRecord,
      }),
    ).toEqual({
      kind: 'rejected',
      reason: 'identity-owned-by-another-account',
    });
  });
});

File: backend/internal/identity/email_otp_test.go
package identity

import (
	"encoding/json"
	"os"
	"strings"
	"testing"
)

const (
	testEmailOtpChallengeID = "01991f20-61d2-7000-8000-000000000501"
	testEmailOtpAddress     = "Person@example.com"
	testEmailOtpSecret      = "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
	testEmailOtpOtherSecret = "BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBE"
)

func TestEmailOtpValueParsersAndNormalization(t *testing.T) {
	address, err := ParseEmailOtpAddress("Person@Example.COM")
	if err != nil || address != testEmailOtpAddress {
		t.Fatalf("ParseEmailOtpAddress() = %q, %v", address, err)
	}
	verified, err := ParseVerifiedEmailAddress("Person@Example.COM")
	if err != nil || verified != VerifiedEmailAddress(testEmailOtpAddress) {
		t.Fatalf("ParseVerifiedEmailAddress() = %q, %v", verified, err)
	}
	oidcEmail, err := ParseOidcEmailAddress("Person@Example.COM")
	if err != nil || oidcEmail.Verified() != verified {
		t.Fatalf("ParseOidcEmailAddress() = %q, %v", oidcEmail, err)
	}
	for _, candidate := range []string{
		"person..x@example.com", ".person@example.com", "person@example.com ",
		"person@-example.com", "person@example..com", "person@example.com@evil.test",
	} {
		if _, err := ParseEmailOtpAddress(candidate); err == nil {
			t.Errorf("ParseEmailOtpAddress(%q) succeeded", candidate)
		}
	}
	if _, err := ParseEmailOtpCode("12345678"); err != nil {
		t.Fatalf("ParseEmailOtpCode() error = %v", err)
	}
	for _, candidate := range []string{"1234567", "123456789", "1234abcd"} {
		if _, err := ParseEmailOtpCode(candidate); err == nil {
			t.Errorf("ParseEmailOtpCode(%q) succeeded", candidate)
		}
	}
	if _, err := ParseEmailOtpChallengeID("not-a-uuid"); err == nil {
		t.Fatal("malformed challenge ID accepted")
	}
	if _, err := ParseEmailOtpDigest("AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB"); err == nil {
		t.Fatal("non-canonical digest accepted")
	}
}

func TestSharedEmailOtpFixture(t *testing.T) {
	t.Parallel()
	content, err := os.ReadFile("../../../contracts/fixtures/identity/email-otp.json")
	if err != nil {
		t.Fatal(err)
	}
	var fixture struct {
		Profile               string `json:"profile"`
		ChallengeID           string `json:"challengeId"`
		Address               string `json:"address"`
		Code                  string `json:"code"`
		Salt                  string `json:"salt"`
		Digest                string `json:"digest"`
		CreatedAtEpochSeconds int64  `json:"createdAtEpochSeconds"`
		VerifyAtEpochSeconds  int64  `json:"verifyAtEpochSeconds"`
		ResendAtEpochSeconds  int64  `json:"resendAtEpochSeconds"`
		Expected              struct {
			CanonicalAddress      string `json:"canonicalAddress"`
			ExpiresAtEpochSeconds int64  `json:"expiresAtEpochSeconds"`
			FailedAttempts        int64  `json:"failedAttempts"`
			SendCount             int64  `json:"sendCount"`
			Version               int64  `json:"version"`
		} `json:"expected"`
	}
	decoder := json.NewDecoder(strings.NewReader(string(content)))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(&fixture); err != nil {
		t.Fatalf("decode shared Email OTP fixture: %v", err)
	}
	address, err := ParseEmailOtpAddress(fixture.Address)
	if err != nil {
		t.Fatal(err)
	}
	created := CreateEmailOtpChallenge(EmailOtpChallenge{
		ChallengeID: mustEmailOtpChallengeID(t, fixture.ChallengeID), Address: address,
		Digest: mustEmailOtpDigest(t, fixture.Digest), Salt: mustEmailOtpSalt(t, fixture.Salt),
		Purpose: EmailOtpPurpose{Kind: EmailOtpPurposeSignIn}, CreatedAtEpochSeconds: fixture.CreatedAtEpochSeconds,
	})
	if fixture.Profile != "email-otp-core" || !created.Created ||
		string(created.Challenge.Address) != fixture.Expected.CanonicalAddress ||
		created.Challenge.ExpiresAtEpochSeconds != fixture.Expected.ExpiresAtEpochSeconds ||
		created.Challenge.FailedAttempts != fixture.Expected.FailedAttempts ||
		created.Challenge.SendCount != fixture.Expected.SendCount || created.Challenge.Version != fixture.Expected.Version {
		t.Fatalf("shared fixture creation = %#v", created)
	}
	if _, err := ParseEmailOtpCode(fixture.Code); err != nil {
		t.Fatal(err)
	}
	verified := VerifyEmailOtpChallenge(created.Challenge, true, fixture.VerifyAtEpochSeconds)
	if !verified.Verified || verified.Challenge.Kind != EmailOtpConsumed {
		t.Fatalf("shared fixture verification = %#v", verified)
	}
	resent := ResendEmailOtpChallenge(
		created.Challenge,
		mustEmailOtpDigest(t, fixture.Digest),
		mustEmailOtpSalt(t, fixture.Salt),
		fixture.ResendAtEpochSeconds,
	)
	if !resent.Resent || resent.Challenge.ExpiresAtEpochSeconds != fixture.Expected.ExpiresAtEpochSeconds ||
		resent.Challenge.FailedAttempts != fixture.Expected.FailedAttempts || resent.Challenge.SendCount != 2 {
		t.Fatalf("shared fixture resend = %#v", resent)
	}
}

func TestEmailOtpChallengeLifecycle(t *testing.T) {
	created := CreateEmailOtpChallenge(emailOtpChallengeFixture(t))
	if !created.Created || !ValidEmailOtpChallenge(created.Challenge) {
		t.Fatalf("CreateEmailOtpChallenge() = %#v", created)
	}
	if created.Challenge.ExpiresAtEpochSeconds != 1_600 || created.Challenge.Version != 1 {
		t.Fatalf("created challenge = %#v", created.Challenge)
	}
	if decision := CreateEmailOtpChallenge(EmailOtpChallenge{
		ChallengeID: emailOtpChallengeID(t), Address: emailOtpAddress(t),
		Digest: emailOtpDigest(t), Salt: emailOtpSalt(t),
		Purpose: EmailOtpPurpose{Kind: EmailOtpPurposeSignIn}, CreatedAtEpochSeconds: MaximumSafeInteger,
	}); decision.Created || decision.Reason != CreateEmailOtpInvalidClock {
		t.Fatalf("unsafe clock decision = %#v", decision)
	}

	verified := VerifyEmailOtpChallenge(created.Challenge, true, 1_500)
	if !verified.Verified || verified.Challenge.Kind != EmailOtpConsumed ||
		!ValidEmailOtpChallenge(verified.Challenge) {
		t.Fatalf("verified decision = %#v", verified)
	}
	if replay := VerifyEmailOtpChallenge(verified.Challenge, true, 1_501); replay.Verified || replay.Reason != VerifyEmailOtpNotPending {
		t.Fatalf("replay decision = %#v", replay)
	}
	expired := VerifyEmailOtpChallenge(created.Challenge, true, 1_600)
	if !expired.Changed || expired.Reason != VerifyEmailOtpExpired ||
		expired.Challenge.InvalidationReason != EmailOtpExpired || !ValidEmailOtpChallenge(expired.Challenge) {
		t.Fatalf("expired decision = %#v", expired)
	}
}

func TestEmailOtpLocksAfterFiveFailures(t *testing.T) {
	challenge := CreateEmailOtpChallenge(emailOtpChallengeFixture(t)).Challenge
	for attempt := int64(1); attempt <= EmailOtpMaximumFailedAttempts; attempt++ {
		decision := VerifyEmailOtpChallenge(challenge, false, 1_000+attempt)
		if decision.Verified || !decision.Changed || decision.Challenge.FailedAttempts != attempt {
			t.Fatalf("attempt %d = %#v", attempt, decision)
		}
		challenge = decision.Challenge
	}
	if challenge.Kind != EmailOtpLocked || !ValidEmailOtpChallenge(challenge) {
		t.Fatalf("locked challenge = %#v", challenge)
	}
	if decision := VerifyEmailOtpChallenge(challenge, true, 1_100); decision.Verified || decision.Reason != VerifyEmailOtpNotPending {
		t.Fatalf("post-lock decision = %#v", decision)
	}
}

func TestEmailOtpResendPreservesFailuresAndExpiry(t *testing.T) {
	challenge := CreateEmailOtpChallenge(emailOtpChallengeFixture(t)).Challenge
	failed := VerifyEmailOtpChallenge(challenge, false, 1_001)
	challenge = failed.Challenge
	tooSoon := ResendEmailOtpChallenge(challenge, emailOtpOtherDigest(t), emailOtpOtherSalt(t), 1_059)
	if tooSoon.Resent || tooSoon.Reason != ResendEmailOtpTooSoon {
		t.Fatalf("too-soon decision = %#v", tooSoon)
	}
	resent := ResendEmailOtpChallenge(challenge, emailOtpOtherDigest(t), emailOtpOtherSalt(t), 1_060)
	if !resent.Resent || resent.Challenge.FailedAttempts != 1 || resent.Challenge.SendCount != 2 ||
		resent.Challenge.ExpiresAtEpochSeconds != 1_600 || resent.Challenge.Version != 3 ||
		!ValidEmailOtpChallenge(resent.Challenge) {
		t.Fatalf("resent decision = %#v", resent)
	}
	third := ResendEmailOtpChallenge(resent.Challenge, emailOtpDigest(t), emailOtpSalt(t), 1_120)
	if !third.Resent {
		t.Fatalf("third send = %#v", third)
	}
	limited := ResendEmailOtpChallenge(third.Challenge, emailOtpOtherDigest(t), emailOtpOtherSalt(t), 1_180)
	if limited.Resent || limited.Reason != ResendEmailOtpSendLimit {
		t.Fatalf("send-limit decision = %#v", limited)
	}
	if third.Challenge.FailedAttempts != 1 || third.Challenge.ExpiresAtEpochSeconds != 1_600 {
		t.Fatalf("resend reset protection = %#v", third.Challenge)
	}
}

func TestEmailOtpDeliveryInvalidation(t *testing.T) {
	challenge := CreateEmailOtpChallenge(emailOtpChallengeFixture(t)).Challenge
	invalidated, ok := InvalidateEmailOtpDelivery(challenge, 1_001)
	if !ok || invalidated.Kind != EmailOtpInvalidated || invalidated.InvalidationReason != EmailOtpDeliveryFailed ||
		!ValidEmailOtpChallenge(invalidated) {
		t.Fatalf("InvalidateEmailOtpDelivery() = %#v, %t", invalidated, ok)
	}
	if _, ok := InvalidateEmailOtpDelivery(challenge, 1_600); ok {
		t.Fatal("delivery failure after expiry was accepted")
	}
}

func TestEmailOtpRateLimitPolicy(t *testing.T) {
	state := EmailOtpRateLimitState{
		Address: emailOtpBucket(t, 4, 1_000, 'A'),
		Network: emailOtpBucket(t, 4, 1_000, 'B'),
	}
	reserved := ReserveEmailOtpRateLimit(state, 1_100)
	if !reserved.Reserved || reserved.State.Address.Count != 5 || reserved.State.Network.Count != 5 {
		t.Fatalf("reserve decision = %#v", reserved)
	}
	state.Address.Count = 5
	if limited := ReserveEmailOtpRateLimit(state, 1_100); limited.Reserved || limited.Reason != "limited" {
		t.Fatalf("limited decision = %#v", limited)
	}
	account := emailOtpBucket(t, 5, 1_000, 'C')
	state.Account = &account
	state.Network.Count = 30
	reset := ReserveEmailOtpRateLimit(state, 1_000+EmailOtpRateLimitWindowSeconds)
	if !reset.Reserved || reset.State.Address.Count != 1 || reset.State.Network.Count != 1 ||
		reset.State.Account == nil || reset.State.Account.Count != 1 {
		t.Fatalf("reset decision = %#v", reset)
	}
	if account.Count != 5 {
		t.Fatalf("caller-owned account bucket mutated: %#v", account)
	}
}

func TestEmailOtpIdentityResolutionNeverAutoLinks(t *testing.T) {
	accountID := mustAccountID(t, "01991f20-61d2-7000-8000-000000000101")
	otherAccountID := mustAccountID(t, "01991f20-61d2-7000-8000-000000000102")
	record := EmailOtpIdentityRecord{
		IdentityID: mustIdentityID(t, "01991f20-61d2-7000-8000-000000000301"),
		AccountID:  accountID, VaultID: mustVaultID(t, "01991f20-61d2-7000-8000-000000000201"),
		Address: emailOtpAddress(t),
	}
	owner := accountID
	collision := DecideEmailOtpIdentityResolution(struct {
		Purpose                     EmailOtpPurpose
		Address                     EmailOtpAddress
		ExistingIdentity            *EmailOtpIdentityRecord
		VerifiedEmailOwnerAccountID *AccountID
	}{Purpose: EmailOtpPurpose{Kind: EmailOtpPurposeSignIn}, Address: emailOtpAddress(t), VerifiedEmailOwnerAccountID: &owner})
	if collision.Kind != EmailOtpIdentityRejected || collision.Reason != EmailOtpIdentityEmailCollision {
		t.Fatalf("collision = %#v", collision)
	}
	authenticated := DecideEmailOtpIdentityResolution(struct {
		Purpose                     EmailOtpPurpose
		Address                     EmailOtpAddress
		ExistingIdentity            *EmailOtpIdentityRecord
		VerifiedEmailOwnerAccountID *AccountID
	}{Purpose: EmailOtpPurpose{Kind: EmailOtpPurposeSignIn}, Address: emailOtpAddress(t), ExistingIdentity: &record})
	if authenticated.Kind != EmailOtpAuthenticateExisting {
		t.Fatalf("authenticated = %#v", authenticated)
	}
	wrongOwner := DecideEmailOtpIdentityResolution(struct {
		Purpose                     EmailOtpPurpose
		Address                     EmailOtpAddress
		ExistingIdentity            *EmailOtpIdentityRecord
		VerifiedEmailOwnerAccountID *AccountID
	}{Purpose: EmailOtpPurpose{Kind: EmailOtpPurposeLink, AccountID: otherAccountID}, Address: emailOtpAddress(t), ExistingIdentity: &record})
	if wrongOwner.Kind != EmailOtpIdentityRejected || wrongOwner.Reason != EmailOtpIdentityOwnedElsewhere {
		t.Fatalf("wrong owner = %#v", wrongOwner)
	}
}

func emailOtpChallengeFixture(t *testing.T) EmailOtpChallenge {
	t.Helper()
	return EmailOtpChallenge{
		ChallengeID: emailOtpChallengeID(t), Address: emailOtpAddress(t),
		Digest: emailOtpDigest(t), Salt: emailOtpSalt(t),
		Purpose: EmailOtpPurpose{Kind: EmailOtpPurposeSignIn}, CreatedAtEpochSeconds: 1_000,
	}
}

func emailOtpChallengeID(t *testing.T) EmailOtpChallengeID {
	t.Helper()
	value, err := ParseEmailOtpChallengeID(testEmailOtpChallengeID)
	if err != nil {
		t.Fatal(err)
	}
	return value
}

func mustEmailOtpChallengeID(t *testing.T, raw string) EmailOtpChallengeID {
	t.Helper()
	value, err := ParseEmailOtpChallengeID(raw)
	if err != nil {
		t.Fatal(err)
	}
	return value
}

func emailOtpAddress(t *testing.T) EmailOtpAddress {
	t.Helper()
	value, err := ParseEmailOtpAddress(testEmailOtpAddress)
	if err != nil {
		t.Fatal(err)
	}
	return value
}

func emailOtpDigest(t *testing.T) EmailOtpDigest {
	t.Helper()
	value, err := ParseEmailOtpDigest(testEmailOtpSecret)
	if err != nil {
		t.Fatal(err)
	}
	return value
}

func mustEmailOtpDigest(t *testing.T, raw string) EmailOtpDigest {
	t.Helper()
	value, err := ParseEmailOtpDigest(raw)
	if err != nil {
		t.Fatal(err)
	}
	return value
}

func emailOtpOtherDigest(t *testing.T) EmailOtpDigest {
	t.Helper()
	value, err := ParseEmailOtpDigest(testEmailOtpOtherSecret)
	if err != nil {
		t.Fatal(err)
	}
	return value
}

func emailOtpSalt(t *testing.T) EmailOtpSalt {
	t.Helper()
	value, err := ParseEmailOtpSalt(testEmailOtpSecret)
	if err != nil {
		t.Fatal(err)
	}
	return value
}

func mustEmailOtpSalt(t *testing.T, raw string) EmailOtpSalt {
	t.Helper()
	value, err := ParseEmailOtpSalt(raw)
	if err != nil {
		t.Fatal(err)
	}
	return value
}

func emailOtpOtherSalt(t *testing.T) EmailOtpSalt {
	t.Helper()
	value, err := ParseEmailOtpSalt(testEmailOtpOtherSecret)
	if err != nil {
		t.Fatal(err)
	}
	return value
}

func emailOtpBucket(t *testing.T, count int64, startedAt int64, repeated byte) EmailOtpRateLimitBucket {
	t.Helper()
	raw := string(make([]byte, 42))
	buffer := []byte(raw)
	for index := range buffer {
		buffer[index] = repeated
	}
	buffer = append(buffer, 'A')
	key, err := ParseEmailOtpRateLimitKey(string(buffer))
	if err != nil {
		t.Fatal(err)
	}
	return EmailOtpRateLimitBucket{Key: key, Count: count, WindowStartedAtEpochSeconds: startedAt}
}
